src/state/slices/opportunitiesSlice/resolvers/foxy/index.ts

```diff
@@ -0,0 +1,172 @@
+    const underlyingAsset = selectAssetById(state, tokenAssetId)
+    const marketData = selectMarketDataById(state, tokenAssetId)
+
+    if (!underlyingAsset) continue
+
+    const tvl = bnOrZero(opportunity.tvl)
+      .div(`1e+${underlyingAsset?.precision}`)
+      .times(marketData.price)
+      .toString()
+
+    const apy = foxyApr.data?.foxyApr ?? '0'
+
+    stakingOpportunitiesById[opportunityId] = {
+      apy,
+      assetId,
+      provider: DefiProvider.ShapeShift as const,
+      tvl,
+      type: DefiType.Staking as const,
+      underlyingAssetId: rewardTokenAssetId,
+      underlyingAssetIds: [tokenAssetId],
+      underlyingAssetRatiosBaseUnit: ['1'],
+      name: `${underlyingAsset.symbol}`,
+    }
+  }
+
+  const data = {
+    byId: stakingOpportunitiesById,
+    type: opportunityType,
+  }
+
+  return { data }
+}
+
+export const foxyStakingOpportunitiesUserDataResolver = async ({
+  opportunityType,
+  accountId,
+  reduxApi,
+  opportunityIds,
+}: OpportunitiesUserDataResolverInput): Promise<{ data: GetOpportunityUserStakingDataOutput }> => {
+  const { getState } = reduxApi
+  const state: any = getState() // ReduxState causes circular dependency
+
+  const stakingOpportunitiesUserDataByUserStakingId: OpportunitiesState['userStaking']['byId'] = {}
+
+  const foxyInvestor = getFoxyApi()
+
+  for (const stakingOpportunityId of opportunityIds) {
+    const balanceFilter = { accountId, assetId: foxyAssetId }
+    const balance = selectPortfolioCryptoBalanceByFilter(state, balanceFilter)
+
+    const asset = selectAssetById(state, foxyAssetId)
+    if (!asset) continue
+
+    const toAssetIdParts: ToAssetIdArgs = {
+      assetNamespace: fromAssetId(stakingOpportunityId).assetNamespace,
+      assetReference: fromAssetId(stakingOpportunityId).assetReference,
+      chainId: fromAssetId(stakingOpportunityId).chainId,
+    }
+    const opportunityId = toOpportunityId(toAssetIdParts)
+    const userStakingId = serializeUserStakingId(accountId, opportunityId)
+
+    // This works because of Idle assets being both a portfolio-owned asset and a yield-bearing "staking asset"
+    // If you use me as a reference and copy me into a resolver for another opportunity, that might or might not be the case
+    // Don't do what monkey see, and adapt the business logic to the opportunity you're implementing
+    if (bnOrZero(balance).eq(0)) {
+      // Zero out this user staking opportunity including rewards - all rewards are automatically claimed when withdrawing, see
+      // https://docs.idle.finance/developers/best-yield/methods/redeemidletoken-1
+      // https://docs.idle.finance/developers/perpetual-yield-tranches/methods/withdrawbb
+      stakingOpportunitiesUserDataByUserStakingId[userStakingId] = {
+        stakedAmountCryptoBaseUnit: '0',
+        rewardsAmountsCryptoBaseUnit: [],
+      }
+      continue
+    }
+
+    const opportunity = await foxyInvestor.getFoxyOpportunities()
```

Comment:
Might be an array of 0 but still
```suggestion
    const opportunities = await foxyInvestor.getFoxyOpportunities()
```

Reply:
Alternatively, you can destructure it like so, since `getFoxyOpportunities`  will not return undefined or 0-length array, but rather throw in case of any error, so if any error is thrown on the right side, the destructuring will never be ran 

```suggestion
    const [opportunity] = await foxyInvestor.getFoxyOpportunities()
```